libbeat/outputs/shipper/shipper.go

```diff
@@ -164,90 +193,135 @@ func (c *shipper) Publish(ctx context.Context, batch publisher.Batch) error {
+		}
 
-	for indexReply.PersistedIndex < publishReply.AcceptedIndex {
-		indexReply, err = ackClient.Recv()
-		if err != nil {
-			return fmt.Errorf("acknowledgement failed due to the connectivity error: %w", err)
+		// with a correct server implementation should never happen, this error is not recoverable
+		if int(publishReply.AcceptedCount) > len(nonDroppedEvents) {
+			return fmt.Errorf(
+				"server returned unexpected results, expected maximum accepted items %d, got %d",
+				len(nonDroppedEvents),
+				publishReply.AcceptedCount,
+			)
 		}
+		toSend = toSend[publishReply.AcceptedCount:]
+		lastAcceptedIndex = publishReply.AcceptedIndex
 	}
 
-	retries := nonDroppedEvents[publishReply.AcceptedCount:]
-	if len(retries) == 0 {
-		batch.ACK()
-		st.Acked(len(nonDroppedEvents))
-		c.log.Debugf("%d events have been accepted, %d dropped", len(nonDroppedEvents), droppedCount)
-	} else {
-		batch.RetryEvents(retries) // decreases TTL unless guaranteed delivery
-		st.Failed(len(retries))
-		c.log.Debugf("%d events have been accepted, %d sent for retry, %d dropped", publishReply.AcceptedCount, len(retries), droppedCount)
-	}
+	s.log.Debugf("%d events have been accepted, %d dropped", len(nonDroppedEvents), droppedCount)
+
+	s.pendingMutex.Lock()
+	s.pending = append(s.pending, pendingBatch{
+		batch:        batch,
+		index:        lastAcceptedIndex,
+		serverID:     s.serverID,
+		droppedCount: droppedCount,
+	})
+	s.pendingMutex.Unlock()
 
 	return nil
 }
 
 // Close closes the connection to the shipper server.
 // Also, implements `outputs.Client`
-func (c *shipper) Close() error {
-	if c.client == nil {
+func (s *shipper) Close() error {
+	if s.client == nil {
 		return fmt.Errorf("connection is not established")
 	}
-	err := c.conn.Close()
-	c.conn = nil
-	c.client = nil
+	s.backgroundCancel()
+	err := s.conn.Close()
+	s.conn = nil
+	s.client = nil
 
 	return err
 }
 
 // String implements `outputs.Client`
-func (c *shipper) String() string {
+func (s *shipper) String() string {
 	return "shipper"
 }
 
+func (s *shipper) ackLoop(ctx context.Context) error {
+	st := s.observer
+
+	// so we explicitly close the stream when exit the function
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	s.clientMutex.Lock()
+	ackClient, err := s.client.PersistedIndex(ctx, &messages.PersistedIndexRequest{
+		PollingInterval: durationpb.New(s.config.AckPollingInterval),
+	})
+	s.clientMutex.Unlock()
+	if err != nil {
+		return fmt.Errorf("acknowledgment failed due to the connectivity error: %w", err)
+	}
+
+	for {
+		select {
+
+		case <-ctx.Done():
+			return ctx.Err()
+
+		default:
+			// this sends an update and unblocks only if the `PersistedIndex` value has changed
+			indexReply, err := ackClient.Recv()
+			if err != nil {
+				return fmt.Errorf("acknowledgment failed due to the connectivity error: %w", err)
+			}
+
+			s.pendingMutex.Lock()
+			lastProcessed := 0
+			for _, p := range s.pending {
+				if p.serverID != indexReply.Uuid {
```

Comment:
Going along with the initialization comment above: the cleanup from a mismatched uuid should happen in `Publish` and/or `Close` -- the uuid will never change during an active connection, so after the first iteration this would just be caught by the `err != nil` check above and none of the outstanding batches would be cancelled. (It might also be nice for readability to move this into a standalone `cancelAllBatches` helper.)

Reply:
See my previous comment https://github.com/elastic/beats/pull/32785#discussion_r954582755

Reply:
That makes sense, but in that case can we move the cleanup so it happens in `Connect` when we set `s.serverID`? It will still never change over the course of a connection, so this check could be skipped -- keeping one-time initialization separate makes the logic of `ackLoop` clearer.

Reply:
It's not going to simplify the rest of the logic, in fact it's going to add one more lock for the pending queue in a different place that can possibly cause a deadlock with the client lock. Also, I would have to copy items in two places. It's more robust to keep it the way it is and I don't see any reason to move it. Unless there is a good argument why the current state of the code does not work as intended I'm going to keep it.